packages/sdk/encrypted-backup/src/backup.ts

```diff
@@ -0,0 +1,431 @@
+    )
+    const unwrap = await circuitBreakerClient.unwrapKey(backup.encryptedFuseKey)
+    if (!unwrap.ok) {
+      return Err(unwrap.error)
+    }
+
+    // Mix the fuse key into the ongoing key hardening. Note that mixing in the circuit breaker key
+    // occurs before the request to ODIS. This means an attacker would need to aquire the fuse key
+    // _before_ they can make any attempts to guess the user's secret.
+    updatedKey = deriveKey(KDFInfo.FUSE_KEY, [initialKey, unwrap.result])
+  } else {
+    debug('backup did not specify an encrypted fuse key')
+    updatedKey = initialKey
+  }
+
+  // If ODIS is in use, harden the key with the output of a rate limited ODIS POPRF function.
+  let odisHardenedKey: Buffer | undefined
+  if (backup.odisDomain !== undefined) {
+    const domain = backup.odisDomain
+
+    // Derive the query authorizer wallet and address from the nonce.
+    // If the ODIS domain is authenticated, the authorizer address should match the domain.
+    const authorizer = odisQueryAuthorizer(backup.nonce)
+    if (domain.address.defined && !eqAddress(authorizer.address, domain.address.value)) {
+      return Err(
+        new InvalidBackupError(
+          new Error(
+            'domain query authorizer address is provided but is not derived from the backup nonce'
+          )
+        )
+      )
+    }
+    if (backup.environment?.odis === undefined) {
+      return Err(
+        new InvalidBackupError(
+          new Error('ODIS domain is provided by no ODIS environment information')
+        )
+      )
+    }
+
+    debug('requesting a key hardening response from ODIS')
+    const odisHardening = await odisHardenKey(
+      updatedKey,
+      domain,
+      backup.environment.odis,
+      authorizer.wallet
+    )
+    if (!odisHardening.ok) {
+      return Err(odisHardening.error)
+    }
+    odisHardenedKey = odisHardening.result
+  } else {
+    debug('not using ODIS for key hardening')
+  }
+
+  let computationalHardenedKey: Buffer | undefined
+  if (backup.computationalHardening !== undefined) {
+    debug('hardening user key with computational function', backup.computationalHardening)
+    const computationalHardening = await computationalHardenKey(
+      updatedKey,
+      backup.computationalHardening
+    )
+    if (!computationalHardening.ok) {
+      return Err(computationalHardening.error)
+    }
+    computationalHardenedKey = computationalHardening.result
+  } else {
+    debug('not using computational key hardening')
+  }
+
+  debug('finalizing decryption key')
+  const finalKey = deriveKey(
+    KDFInfo.FINALIZE,
+    [updatedKey, odisHardenedKey ?? Buffer.alloc(0), computationalHardenedKey ?? Buffer.alloc(0)],
+    16
+  )
+
+  debug('decrypting backup with finalized decryption key')
+  const decryption = decrypt(finalKey, backup.encryptedData)
+  if (!decryption.ok) {
+    return Err(decryption.error)
+  }
+
+  debug('requesting a key hardening response from ODIS')
```

Comment:
I think this is repeated by mistake

Reply:
Changed the message in 37a1137b6c640cbc64caa0b165431e331d047242